### DISCUSSION TITLE
CAPD Infrastructure Provider

### DISCUSSION THREAD

**Discussion by Sir-gits-a-lot** [kubernetes-sigs/cluster-api#8696] [General]
cluster api not able to reconcile for docker infrastructure provider, I have created a k8s cluster using kind. 

Initialized it using clusterctl init --infrastructure docker and apply one CAPI manifest for 3 control-node + 3 worker node. It was successfully deployed, all the machines were up (docker containers). 

After that I am trying to test the benefits of doing this whole thing using CAPI (Reconcillation, full lifecycle management of workload clusters). 

For that I stopped 1 container(worker node machine), So even though Cluster CAPI detects the change and starts the reconcillation process (this I can say by checking the capd-controller-manager pod logs, it was not able to set up the worker node. I expected it to launch another conatiner and add it to the cluster but it's not able to do that. 

There is no issue with memory, I have increased the inotify limit. 

How I can resolve this?

**Comment by killianmuldoon** [comment]
Normally we use issues for this sort of thing - could you close this discussion and move it into an issue?

In general Cluster API won't try to replace a machine when it becomes unhealthy unless you have MachineHealthChecks (MHCs) enabled. I don't believe these are part of the quickstart version of the ClusterClass. You could try enabling MHCs to see if that gives you the expected behaviour.

**Reply by Sir-gits-a-lot** [reply]
Sure, I'll move it to issue. Please check and respond if you feel like.

**Reply by Sir-gits-a-lot** [reply]
Actually, I thought that since initializing CAPI (Making current kind cluster into management cluster) would involve making all the CRDs (Like K8s clusters as a CRD) and Custom Resource Controllers (for the K8s clusters CRD) so, it make sense that all the features of K8s like reconciliation (full lifecycle management of workload clusters) will automatically be handled by the management cluster from now onwards. Why do we have to define MHCs, then what does the process of transforming a normal K8s cluster into CAPI Management Cluster actually do? Because surely it doesn't create CR controllers for the specific providers as we have to do it using MHCs.

Is my understanding wrong in this case?

**Reply by Sir-gits-a-lot** [reply]
I am trying to enable the machineHealthCheck, but I think it's not supported for docker provider. I am not finding a conclusive statement for that though in this page https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/healthchecking.html. 
What I don't understand is should CAPI take care of the recociliation itself, there should be custom resource controller to do the requisite reconciliation. That was the whole selling point of using K8s api's to manage K8s, right?

**Comment by killianmuldoon** [comment]
Let's continue on the issue.
